### PR TITLE
Add mithril jsx types

### PR DESCRIPTION
**Pull request by fent**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L2823-L3027 <- I basically copied the types from here
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

Related to https://github.com/Microsoft/TypeScript/issues/3679

solutions provided there are to install react types, `@types/react`, which contain the interfaces jsx looks for. but if using something other than react for jsx, types collide.

### CODE REVIEW

**Review by dead-claudia — CHANGES_REQUESTED** [review]
Have you tested this locally to ensure it works? Otherwise, LGTM.

**Comment by fent** [comment]
> Have you tested this locally to ensure it works? Otherwise, LGTM.

I did. I symlinked my fork to a project I'm working on that uses jsx with mithril. Not getting any warnings after.

**Review by dead-claudia — APPROVED** [review]

**Comment by isiahmeadows** [comment]
That's awesome! Thanks @fent for the addition.

**Comment by spacejack** [comment]
This looks good to me. @andraaspar you might want to take a look, I'm not sure if this will conflict with your [userland solution](https://github.com/andraaspar/mithril-tsx-component)? These types do not appear to need any JS implementation.

I think it would be nice to add a test, eg:

```tsx
// test-tsx.tsx
import * as m from 'mithril'

interface Attrs {
	id: number;
	text: string;
}

const PojoComp: m.Component<Attrs> = {
	view: ({attrs}) => <p class="abc">{attrs.text}</p>
};

class ClassComp implements m.ClassComponent<Attrs> {
	view({attrs}: m.Vnode<Attrs>) {
		return <p class="abc">{attrs.text}</p>;
	}
}

function ClosureComp(): m.Component<Attrs> {
	return {
		view: ({attrs}) => <p class="abc">{attrs.text}</p>
	};
}

function testRender() {
	return [
		<ClassComp id={1} text="Title 1"/>,
		// <PojoComp id={2} text="Title 2"/>,  <- won't compile
		// <ClosureComp id={3} text="Title 3"/>  <- won't compile
	];
}
```
Unfortunately Typescript/TSX does not seem to be able to use POJO or Closure components yet. I don't know if there is a solution for this.

File: types/mithril/index.d.ts [hunk 1]
```diff
@@ -294,5 +294,197 @@ declare namespace Mithril {
...
+    namespace JSX {
+        // tslint:disable-next-line:no-empty-interface
+        interface Element extends Mithril.Vnode { }
+
+        // tslint:disable-next-line:no-empty-interface
+        interface IntrinsicAttributes extends Mithril.Attributes { }
+        // tslint:disable-next-line:no-empty-interface
+        interface IntrinsicClassAttributes extends Mithril.Attributes { }
+
+        interface IntrinsicElements {
```
Comment (isiahmeadows) [inline types/mithril/index.d.ts]: Forgot to mention: please include the various special vnode types (especially `"["` for fragments) in this list. Those are pretty important.

**Review by dead-claudia — CHANGES_REQUESTED** [review]

**Comment by isiahmeadows** [comment]
Also, I'd like to invite you to participate in https://github.com/MithrilJS/mithril.d.ts for further discussion on this topic. A few issues were raised in the Gitter chat room, and I'd like this to be discussed at least somewhat.

File: types/mithril/index.d.ts (hunk 1 above)
Comment (fent) [reply types/mithril/index.d.ts]: how can I do this? adding "[" as a key and pointing to `Mithril.Attributes`? what are the other special vnode types?

Comment (isiahmeadows) [reply types/mithril/index.d.ts]: Yes. You literally add those to the list.

**Review by dead-claudia — APPROVED** [review]
LGTM

**Pull request merged**